core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -28,41 +34,180 @@
  * @author iyzhang
  */
 public class KernelClient {
+    private static final int MIN_METRIC_POLL_PERIOD_MS = 1000; /* Minimum metrics poll period */
+    private static final int MAX_METRIC_POLL_PERIOD_MS = 512000; /* Maximum metrics poll period */
+    private static final int STEP_SIZE_IN_BYTES = 1024; /* Step size in bytes */
+    /* Random byte array used to send in heartbeats */
+    private static final byte[] randomBytes = new byte[STEP_SIZE_IN_BYTES];
+
+    static {
+        new Random().nextBytes(randomBytes);
+    }
+
+    /* Data holder class used in heartbeats */
+    public static class RandomData implements Serializable {
+        private transient int len = STEP_SIZE_IN_BYTES; /* length of data to be sent to server */
+
+        private void writeObject(java.io.ObjectOutputStream s) throws java.io.IOException {
+            s.writeInt(len);
+            for (int i = 0; i < len; i += STEP_SIZE_IN_BYTES) {
+                /* Write the same byte array again as this data is useless on receiver end */
+                s.write(randomBytes);
+            }
+        }
+
+        private void readObject(java.io.ObjectInputStream s)
+                throws java.io.IOException, ClassNotFoundException {
+            len = s.readInt();
+            for (int i = 0; i < len; i += STEP_SIZE_IN_BYTES) {
+                /* Received data is useless. Just overwrites to same byte array */
+                s.read(randomBytes);
+            }
+        }
+    }
+
     /** Stub for the OMS */
     private OMSServer oms;
-    /** List of hostnames matched to kernel server stubs */
-    private Hashtable<InetSocketAddress, KernelServer> servers;
+
+    /* Maximum number of continuous heartbeat miss allowed */
```

Comment:
nit: I think you mean 'consecutive' not 'continuous'?

Reply:
Yeah should be consecutive. Modified it.